packages/lesswrong/components/ultraFeed/UsersNameWithModal.tsx

```diff
@@ -0,0 +1,116 @@
+import React from 'react';
+import UsersName from '../users/UsersName';
+import UltraFeedUserCard from '../ultraFeed/UltraFeedUserCard';
+import UltraFeedUserDialog from './UltraFeedUserDialog';
+import HoverOver from '../common/HoverOver';
+import { userGetDisplayName, userGetProfileUrl } from '../../lib/collections/users/helpers';
+import { Link } from '../../lib/reactRouterWrapper';
+import { AnalyticsContext, useTracking } from '../../lib/analyticsEvents';
+import type { Placement as PopperPlacementType } from "popper.js";
+import UserNameDeleted from '../users/UserNameDeleted';
+import { defineStyles, useStyles } from '../hooks/useStyles';
+import { useDialog } from '../common/withDialog';
+import { useUltraFeedContext } from './UltraFeedContextProvider';
+
+const styles = defineStyles("UsersNameWithModal", () => ({
+  cardWrapper: {
+    padding: 0,
+    background: "unset",
+    maxWidth: "none",
+  },
+}));
+
+/**
+ * A username component specifically for UltraFeed that shows an enhanced card on hover (desktop)
+ * or opens a modal dialog (mobile) when clicked.
+ */
+const UsersNameWithModal = ({
+  user,
+  documentId,
+  nofollow = false,
+  tooltipPlacement = "bottom-start",
+  className,
+  simple = false,
+  ...otherProps
+}: {
+  user?: UsersMinimumInfo | null | undefined;
+  documentId?: string;
+  nofollow?: boolean;
+  tooltipPlacement?: PopperPlacementType;
+  className?: string;
+  simple?: boolean;
+  color?: boolean;
+  pageSectionContext?: string;
+}) => {
+  const classes = useStyles(styles);
+  const { captureEvent } = useTracking();
+  const { openDialog } = useDialog();
+  const { openInNewTab } = useUltraFeedContext();
+
+  if (!user && !documentId) {
+    return <UsersName user={user} documentId={documentId} {...otherProps} className={className} />;
+  }
+
+  if (!user && documentId) {
+    return <UsersName documentId={documentId} {...otherProps} className={className} />;
+  }
+
+  if (!user || user.deleted) {
```

Comment:
I don't think `user` can be missing in this block anymore (given the two previous `if` conditions)

Reply:
I think you're right but the type check isn't that smart